From 742198844d4b951dea09d141ff36a8ebdfbd82f6 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Wed, 24 Apr 2024 10:47:37 -0500
Subject: [PATCH 1/3] center patch

center floating windows recently moved to another monitor
center new windows that are floating or are in floating layout
add fn to center floating windows
---
 config.def.h |  2 ++
 dwl.c        | 44 ++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 46 insertions(+)

diff --git a/config.def.h b/config.def.h
index 22d2171..f5225d9 100644
--- a/config.def.h
+++ b/config.def.h
@@ -13,6 +13,7 @@ static const float focuscolor[]            = COLOR(0x005577ff);
 static const float urgentcolor[]           = COLOR(0xff0000ff);
 /* This conforms to the xdg-protocol. Set the alpha to zero to restore the old behavior */
 static const float fullscreen_bg[]         = {0.1f, 0.1f, 0.1f, 1.0f}; /* You can also use glsl colors */
+static const int respect_monitor_reserved_area = 0;  /* 1 to monitor center while respecting the monitor's reserved area, 0 to monitor center */
 
 /* tagging - TAGCOUNT must be no greater than 31 */
 #define TAGCOUNT (9)
@@ -142,6 +143,7 @@ static const Key keys[] = {
 	{ MODKEY,                    XKB_KEY_space,      setlayout,      {0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_space,      togglefloating, {0} },
 	{ MODKEY,                    XKB_KEY_e,         togglefullscreen, {0} },
+	{ MODKEY,                    XKB_KEY_x,          movecenter,     {0} },
 	{ MODKEY,                    XKB_KEY_0,          view,           {.ui = ~0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright, tag,            {.ui = ~0} },
 	{ MODKEY,                    XKB_KEY_comma,      focusmon,       {.i = WLR_DIRECTION_LEFT} },
diff --git a/dwl.c b/dwl.c
index 145fd01..747ead9 100644
--- a/dwl.c
+++ b/dwl.c
@@ -336,6 +336,8 @@ static void tagmon(const Arg *arg);
 static void tile(Monitor *m);
 static void togglefloating(const Arg *arg);
 static void togglefullscreen(const Arg *arg);
+static void _movecenter(Client *c, int interact);
+static void movecenter(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
 static void unlocksession(struct wl_listener *listener, void *data);
@@ -1641,6 +1643,7 @@ mapnotify(struct wl_listener *listener, void *data)
 	Client *p, *w, *c = wl_container_of(listener, c, map);
 	Monitor *m;
 	int i;
+	struct wlr_box b;
 
 	/* Create scene tree for this client and its border */
 	c->scene = client_surface(c)->data = wlr_scene_tree_create(layers[LyrTile]);
@@ -1689,6 +1692,14 @@ mapnotify(struct wl_listener *listener, void *data)
 	} else {
 		applyrules(c);
 	}
+
+	b = respect_monitor_reserved_area ? c->mon->w : c->mon->m;
+	if (c->isfloating || !c->mon->lt[c->mon->sellt]->arrange) {
+		/* client is floating or in floating layout */
+		c->geom.x = (b.width - c->geom.width) / 2 + b.x;
+		c->geom.y = (b.height - c->geom.height) / 2 + b.y;
+	}
+
 	printstatus();
 
 unset_fullscreen:
@@ -2347,6 +2358,10 @@ setmon(Client *c, Monitor *m, uint32_t newtags)
 		c->tags = newtags ? newtags : m->tagset[m->seltags]; /* assign tags of target monitor */
 		setfullscreen(c, c->isfullscreen); /* This will call arrange(c->mon) */
 		setfloating(c, c->isfloating);
+		if (c->isfloating || !c->mon->lt[c->mon->sellt]->arrange) {
+			/* client is floating or in floating layout */
+			_movecenter(c, 0);
+		}
 	}
 	focusclient(focustop(selmon), 1);
 }
@@ -2683,6 +2698,35 @@ togglefullscreen(const Arg *arg)
 		setfullscreen(sel, !sel->isfullscreen);
 }
 
+void
+_movecenter(Client *c, int interact)
+{
+	struct wlr_box b;
+
+	if (!c || !c->mon) {
+		return;
+	}
+
+	if (!c->isfloating) {
+		return;
+	}
+
+	b = respect_monitor_reserved_area ? c->mon->w : c->mon->m;
+	resize(c, (struct wlr_box){
+		.x = (b.width - c->geom.width) / 2 + b.x,
+		.y = (b.height - c->geom.height) / 2 + b.y,
+		.width = c->geom.width,
+		.height = c->geom.height,
+	}, interact);
+}
+
+void
+movecenter(const Arg *arg)
+{
+	Client *c = focustop(selmon);
+	_movecenter(c, 1);
+}
+
 void
 toggletag(const Arg *arg)
 {
-- 
2.45.2


From be061cf0e2da09975f79740b3104c4b211c90995 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Sat, 20 Jul 2024 10:21:09 -0500
Subject: [PATCH 2/3] f

---
 config.def.h |  9 +++++++++
 dwl.c        | 29 +++++++++++++++++++++++++++++
 2 files changed, 38 insertions(+)

diff --git a/config.def.h b/config.def.h
index f5225d9..3d89fed 100644
--- a/config.def.h
+++ b/config.def.h
@@ -37,6 +37,13 @@ static const Layout layouts[] = {
 	{ "[M]",      monocle },
 };
 
+/* size(s) */
+static const Size sizes[] = {
+  /* width   height */
+	{ 1000,    0.75f },
+	{ 500,     0.5f },
+};
+
 /* monitors */
 /* (x=-1, y=-1) is reserved as an "autoconfigure" monitor position indicator
  * WARNING: negative values other than (-1, -1) cause problems with Xwayland clients
@@ -141,6 +148,8 @@ static const Key keys[] = {
 	{ MODKEY,                    XKB_KEY_f,          setlayout,      {.v = &layouts[1]} },
 	{ MODKEY,                    XKB_KEY_m,          setlayout,      {.v = &layouts[2]} },
 	{ MODKEY,                    XKB_KEY_space,      setlayout,      {0} },
+	{ MODKEY,                    XKB_KEY_b,          setsize,        {.v = &sizes[0]} },
+	{ MODKEY,                    XKB_KEY_n,          setsize,        {.v = &sizes[1]} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_space,      togglefloating, {0} },
 	{ MODKEY,                    XKB_KEY_e,         togglefullscreen, {0} },
 	{ MODKEY,                    XKB_KEY_x,          movecenter,     {0} },
diff --git a/dwl.c b/dwl.c
index 747ead9..e943b1a 100644
--- a/dwl.c
+++ b/dwl.c
@@ -185,6 +185,11 @@ typedef struct {
 	void (*arrange)(Monitor *);
 } Layout;
 
+typedef struct {
+	float newwidth;
+	float newheight;
+} Size;
+
 struct Monitor {
 	struct wl_list link;
 	struct wlr_output *wlr_output;
@@ -324,6 +329,7 @@ static void setfloating(Client *c, int floating);
 static void setfullscreen(Client *c, int fullscreen);
 static void setgamma(struct wl_listener *listener, void *data);
 static void setlayout(const Arg *arg);
+static void setsize(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setmon(Client *c, Monitor *m, uint32_t newtags);
 static void setpsel(struct wl_listener *listener, void *data);
@@ -2324,6 +2330,29 @@ setlayout(const Arg *arg)
 	printstatus();
 }
 
+void
+setsize(const Arg *arg)
+{
+	const Size *size = (const Size *)arg->v;
+	Client *c = focustop(selmon);
+	struct wlr_box b;
+	int newwidth;
+	int newheight;
+
+	if (!selmon || !arg || !arg->v || !c || !c->mon || !c->isfloating)
+		return;
+
+	b = respect_monitor_reserved_area ? c->mon->w : c->mon->m;
+	newwidth = (int)round(size->newwidth ? (size->newwidth <= 1 ? b.width * size->newwidth : size->newwidth) : c->geom.width);
+	newheight = (int)round(size->newheight ? (size->newheight <= 1 ? b.height * size->newheight : size->newheight) : c->geom.height);
+	resize(c, (struct wlr_box){
+		.x = (b.width - newwidth) / 2 + b.x,
+		.y = (b.height - newheight) / 2 + b.y,
+		.width = newwidth,
+		.height = newheight,
+	}, 1);
+}
+
 /* arg > 1.0 will set mfact absolutely */
 void
 setmfact(const Arg *arg)
-- 
2.45.2


From 0e7e555a8b05064f047d0426b36b8768713f5179 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Sun, 28 Jul 2024 19:30:54 -0500
Subject: [PATCH 3/3] f

---
 dwl.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dwl.c b/dwl.c
index e943b1a..61c5c82 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1702,8 +1702,8 @@ mapnotify(struct wl_listener *listener, void *data)
 	b = respect_monitor_reserved_area ? c->mon->w : c->mon->m;
 	if (c->isfloating || !c->mon->lt[c->mon->sellt]->arrange) {
 		/* client is floating or in floating layout */
-		c->geom.x = (b.width - c->geom.width) / 2 + b.x;
-		c->geom.y = (b.height - c->geom.height) / 2 + b.y;
+		c->geom.x = c->geom.x == 0 ? (b.width - c->geom.width) / 2 + b.x : c->geom.x;
+		c->geom.y = c->geom.y == 0 ? (b.height - c->geom.height) / 2 + b.y : c->geom.y;
 	}
 
 	printstatus();
-- 
2.45.2

